### PR TITLE
8302017: Allocate BadPaddingException only if it will be thrown

### PR DESCRIPTION
**Pull request by swesonga**
Backport [334b977259930368160db705c1f2feda0b0e8707](https://github.com/openjdk/jdk17u-dev/commit/300ed9156581adb09e8258bc2da21f3bff4ecbde) from https://github.com/openjdk/jdk21u/pull/79

Passes tier1 tests - tested on Linux x64.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] [JDK-8302017](https://bugs.openjdk.org/browse/JDK-8302017) needs maintainer approval

### Issue
 * [JDK-8302017](https://bugs.openjdk.org/browse/JDK-8302017): Allocate BadPaddingException only if it will be thrown (**Enhancement** - P3 - Approved)

### Reviewers
 * [Brian Stafford](https://openjdk.org/census#bstafford) (@brianjstafford - no project role) ⚠️ Review applies to [0dfde364](https://git.openjdk.org/jdk17u-dev/pull/1864/files/0dfde364cf6f08a7649b10e0c6350b6ee81d8855)
 * [Martin Balao](https://openjdk.org/census#mbalao) (@martinuy - **Reviewer**) ⚠️ Review applies to [aa9d466a](https://git.openjdk.org/jdk17u-dev/pull/1864/files/aa9d466afbd67ae089f42bc549d0635ff7205fd2)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk17u-dev.git pull/1864/head:pull/1864` \
`$ git checkout pull/1864`

Update a local copy of the PR: \
`$ git checkout pull/1864` \
`$ git pull https://git.openjdk.org/jdk17u-dev.git pull/1864/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1864`

View PR using the GUI difftool: \
`$ git pr show -t 1864`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk17u-dev/pull/1864.diff">https://git.openjdk.org/jdk17u-dev/pull/1864.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk17u-dev/pull/1864#issuecomment-1754239712)

### CODE REVIEW

File: src/java.base/share/classes/sun/security/rsa/RSAPadding.java [hunk 1]
```diff
@@ -68,7 +67,7 @@
  * The algorithms (representations) are forwards-compatible: that is,
  * the algorithm described in previous releases are in later releases.
  * However, additional comments/checks/clarifications were added to the
- * later versions based on real-world experience (e.g. stricter v1.5
+ * latter versions based on real-world experience (e.g. stricter v1.5
```
Comment (brianjstafford) [inline src/java.base/share/classes/sun/security/rsa/RSAPadding.java:70]: nit: I'm not sure this change in wording is necessary

Comment (brianjstafford) [reply src/java.base/share/classes/sun/security/rsa/RSAPadding.java:70]: Ah, this comes from the change upstream in tip (I was comparing to the jdk21u backport).

File: src/java.base/share/classes/sun/security/rsa/RSAPadding.java [hunk 2]
```diff
@@ -236,24 +234,22 @@ public int getMaxDataSize() {
...
+    public byte[] pad(byte[] data, int ofs, int len) {
         if (len > maxDataSize) {
-            throw new BadPaddingException("Data must be shorter than "
-                + (maxDataSize + 1) + " bytes but received "
-                + len + " bytes.");
+            return null;
         }
         switch (type) {
         case PAD_NONE:
+            // assert len == paddedSize and data.length - ofs > len?
```
Comment (brianjstafford) [inline src/java.base/share/classes/sun/security/rsa/RSAPadding.java:252]: I think this assert is only relevant on tip, where the output of this function is passed to Output.pass(). However, it's in the jdk21u backport, so I guess we should leave it in for consistency....

File: src/java.base/share/classes/sun/security/rsa/RSASignature.java [hunk 3]
```diff
@@ -257,15 +259,17 @@ public static byte[] decodeSignature(ObjectIdentifier oid, byte[] sig)
...
+        if (!algId.getOID().equals(oid)) {
             throw new IOException("ObjectIdentifier mismatch: "
                 + algId.getOID());
         }
         if (algId.getEncodedParams() != null) {
             throw new IOException("Unexpected AlgorithmId parameters");
         }
-        byte[] digest = values[1].getOctetString();
-        return digest;
+        if (values[1].isConstructed()) {
```
Comment (brianjstafford) [inline src/java.base/share/classes/sun/security/rsa/RSASignature.java:269]: What is the source of these changes? I'm having trouble finding them in the jdk21u backport or the original change in tip.

Comment (swesonga) [reply src/java.base/share/classes/sun/security/rsa/RSASignature.java:269]: This change is from https://github.com/openjdk/jdk21u/commit/14e7d911997d33eba2893991fa0e2f507aa977f8 and should not have been included when resolving the conflicts.

File: src/java.base/share/classes/sun/security/rsa/RSAPadding.java (hunk 1 above)
Comment (swesonga) [reply src/java.base/share/classes/sun/security/rsa/RSAPadding.java:70]: This change is from https://github.com/openjdk/jdk21u/commit/4cec141a90bc5d3b8ec17c024291d9c74a112cd4 and should not have been included when resolving the conflicts.

File: src/java.base/share/classes/sun/security/rsa/RSAPadding.java [hunk 4]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2003, 2020, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2003, 2022, Oracle and/or its affiliates. All rights reserved.
```
Comment (macarte) [inline src/java.base/share/classes/sun/security/rsa/RSAPadding.java:2]: this copyright would need to be 2023 as this is when we're modifying this file

File: src/java.base/share/classes/sun/security/rsa/RSAPadding.java (hunk 2 above)
Comment (brianjstafford) [reply src/java.base/share/classes/sun/security/rsa/RSAPadding.java:253]: I can't seem to find the Output.pass() reference I was looking at previously. I would leave this comment in for consistency with the tip and 21u PRs.

**Review by brianjstafford — APPROVED** [review]
LGTM

**Comment by GoeLin** [comment]
Hi @brianjstafford, do you have the possibility to do further testing? Tier1 and GHA are only a basic sanity check. This change is larger, in a delicate part of the JDK and needed changes, so I think further testing is appropriate.

**Comment by brianjstafford** [comment]
> Hi @brianjstafford, do you have the possibility to do further testing? Tier1 and GHA are only a basic sanity check. This change is larger, in a delicate part of the JDK and needed changes, so I think further testing is appropriate.

For sure - we'll do some additional testing and advise of the results.

**Comment by GoeLin** [comment]
Hi @brianjstafford, I remove the fix-request label in the meantime. Please label again once this is ready.

**Comment by seanjmullan** [comment]
Hi, there is a regression related to this fix that we are investigating (https://bugs.openjdk.org/browse/JDK-8320597). I suggest you hold off on integrating this until we have a better handle on the issue and potential fix.

**Review by martinuy — APPROVED** [review]
Hi @swesonga ,

Thanks for proposing this backport. Only one minor comment: the copyright date change in RSAPadding.java is not needed. In any case, looks good to me.

**Comment by seanjmullan** [comment]
> Hi, there is a regression related to this fix that we are investigating (https://bugs.openjdk.org/browse/JDK-8320597). I suggest you hold off on integrating this until we have a better handle on the issue and potential fix.

The fix for JDK-8320597 is now integrated into the mainline. You should also backport this fix either separately or as part of this fix to ensure you don't introduce the regression. This applies to all OpenJDK backports of this issue.

**Comment by martinuy** [comment]
Thanks @seanjmullan for the heads up. We are planning to integrate both bug separately (JDK-8320597 is in [PR 2052](https://github.com/openjdk/jdk17u-dev/pull/2052))

**Comment by martinuy** [comment]
@swesonga I'll sponsor you. Would you mind fixing the RSAPadding.java copyright date?

**Comment by martinuy** [comment]
@swesonga Can you please run the "integrate" command again so the Skara bot re-adds the "sponsor" tag?

**Pull request closed without merging**